draft-ietf-quic-transport.md

```diff
@@ -550,37 +550,45 @@ of protection against off-path attackers.
 ### Initial Packet {#packet-initial}
 
 The Initial packet uses long headers with a type value of 0x7F.  It carries the
-first cryptographic handshake message sent by the client as well as the
-cryptographic messages sent by the server to perform key exchange. The Initial
-packet is protected by Initial keys as described in {{QUIC-TLS}}.
-
-When an an Initial packet is sent by a client which has not previously received
-a Retry packet from the server, it populates the Destination Connection ID field
-with a randomly selected value.  This MUST be at least 8 octets in length. Until
-a packet is received from the server, the client MUST use the same random value
-unless it also changes the Source Connection ID (which effectively starts a new
-connection attempt).  The randomized Destination Connection ID is used to
-determine packet protection keys.
-
-A server sends its first Initial packet in response to a Client Initial.  A
-server may send multiple Initial packets.  The Cryptographic key exchange could
-require multiple round trips or we might perform retransmissions of this data.
-
-The Destination Connection ID field in the server's Initial packet contains a
-connection ID that is chosen by the recipient of the packet; the Source
-Connection ID includes the connection ID that the sender of the packet wishes to
-use (see {{connection-id}}).
-
-On receiving the first Server Initial packet the client uses the Source
-Connection ID supplied by the server as the Destination Connection ID for
-subsequent packets.
-
-If the client received a Retry packet from the server and is sending a second
-Initial packet, then it sets the Destination Connection ID to the value from the
-Source Connection ID in the Retry packet.
-
-Changing Destination Connection ID also results in a change to the keys used to
-protect the Initial packet.
+first CRYPTO_HS frames sent by the client and the server.
```

Comment:
@marten-seemann why would you run a client and server on the same port? They may both be QUIC, but client and server are fundamentally two different types of services, and should be on separate ports.

Reply:
This is useful in p2p applications. We're planning to do this in IPFS.

There are two main advantages to doing this:
1. It makes NAT traversal easier.
2. It's a shortcut for peer discovery: If one peer receives a QUIC connection from another peer, it knows that this peer is able to accept QUIC connections on the same port, and can announce this to other peers in the network.

Reply:
> But it makes me really sad to lose 1:1 mapping, especially when the numbering can be changed to directly represent the encryption level (i.e. 0 to 3).

Note that there is a possibility of losing 1:1 mapping anyways if we decide to use different type numbers for long header packets with the "payload_length" field. See the discussion on ttps://github.com/quicwg/base-drafts/pull/1301 (intentionally omitted the first letter of the scheme to prevent back-link being created).

Reply:
I'm not sure why it makes NAT traversal easier. I don't know much about IPFS, but ICE-based NAT traversal already breaks the symmetry between the peers, so you know who is who.

Reply:
I think I prefer separate packet types as well.